src/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/PerformanceData/PerfProviderCollection.cs

```diff
@@ -0,0 +1,142 @@
+        private static object s_hiddenInternalSyncObject;
+        private static List<PerfProvider> s_providerList = new List<PerfProvider>();
+        private static Dictionary<object, int> s_counterSetList = new Dictionary<object, int>();
+        private static CounterType[] s_counterTypes = (CounterType[])Enum.GetValues(typeof(CounterType));
+        private static CounterSetInstanceType[] s_counterSetInstanceTypes = (CounterSetInstanceType[])Enum.GetValues(typeof(CounterSetInstanceType));
+
+        private static object s_lockObject
+        {
+            get
+            {
+                if (s_hiddenInternalSyncObject == null)
+                {
+                    object o = new object();
+                    Interlocked.CompareExchange(ref s_hiddenInternalSyncObject, o, null);
+                }
+                return s_hiddenInternalSyncObject;
+            }
+        }
+        
+        internal static PerfProvider QueryProvider(Guid providerGuid)
+        {
+            // Most of the cases should be that the application contains 1 provider that supports several CounterSets;
+            // that is, ContainsKey should succeed except for the first time.            
+            lock (s_lockObject)
+            {
+                foreach (PerfProvider ProviderEntry in s_providerList)
+                {
+                    if (ProviderEntry.m_providerGuid == providerGuid)
+                    {
+                        return ProviderEntry;
+                    }
+                }
+
+                PerfProvider NewProvider = new PerfProvider(providerGuid);
+                s_providerList.Add(NewProvider);
+                return NewProvider;
+            }
+        }
+        
+        internal static void RemoveProvider(Guid providerGuid)
+        {
+            lock (s_lockObject)
+            {
+                PerfProvider MatchedProvider = null;
+
+                foreach (PerfProvider ProviderEntry in s_providerList)
+                {
+                    if (ProviderEntry.m_providerGuid == providerGuid)
+                    {
+                        MatchedProvider = ProviderEntry;
+                    }
+                }
+                if (MatchedProvider != null)
+                {
+                    MatchedProvider.m_hProvider.Dispose();
+                    s_providerList.Remove(MatchedProvider);
+                }
+            }
+        }
+
+        [SuppressMessage("Microsoft.Usage", "CA2208:InstantiateArgumentExceptionsCorrectly")]
+        internal static void RegisterCounterSet(Guid counterSetGuid)
+        {
+            // Input counterSetGuid should not be registered yet. That is, ContainsKey() should fail most of times.
+            lock (s_lockObject)
+            {
+                if (s_counterSetList.ContainsKey(counterSetGuid))
+                {
+                    throw new ArgumentException(SR.Format(SR.Perflib_Argument_CounterSetAlreadyRegister, counterSetGuid), "CounterSetGuid");
```

Comment:
Please change argument name strings to nameof eg `CounterSetGuid` to `nameof(CounterSetGuid)`

Reply:
sorry missed that...i'll do another pass

Reply:
fixed i did a pass and the other should not be method arguments.